pkg/apis/networking/types.go

```diff
@@ -544,3 +544,68 @@ type ServiceBackendPort struct {
 	// +optional
 	Number int32
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterCIDRConfig is the Schema for the clustercidrconfigs API.
+type ClusterCIDRConfig struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   ClusterCIDRConfigSpec
+	Status ClusterCIDRConfigStatus
+}
+
+// ClusterCIDRConfigSpec defines the desired state of ClusterCIDRConfig.
+type ClusterCIDRConfigSpec struct {
+	// NodeSelector defines which nodes the config is applicable to.
+	// An empty or nil NodeSelector functions as a default that applies to all nodes.
+	// This field is immutable.
+	// +optional
+	NodeSelector *api.NodeSelector
+
+	// IPv4 defines the IPv4 CIDR and the PerNodeMaskSize.
+	// At least one of IPv4 or IPv6 must be provided. If both are
+	// provided, the number of IPs allocated to each must be the same
+	// (32 - ipv4.perNodeMaskSize).
+	// This field is immutable.
+	// +optional
+	IPv4 *CIDRConfig
+
+	// IPv6 defines the IPv4 CIDR and the PerNodeMaskSize.
+	// At least one of IPv4 or IPv6 must be provided. If both are
+	// provided, the number of IPs allocated to each must be the same
+	// (128 - ipv6.perNodeMaskSize).
+	// This field is immutable.
+	// +optional
+	IPv6 *CIDRConfig
+}
+
+// CIDRConfig defines the CIDR and Mask size per IP Family(IPv4/IPv6).
+type CIDRConfig struct {
+	// An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64").
+	CIDR string
+
+	// PerNodeMaskSize is the mask size for node cidr.
+	// IPv4/IPv6 Netmask size (e.g. 25 -> "/25" or 112 -> "/112") to allocate to a node.
+	PerNodeMaskSize int32
```

Comment:
Bringing back an old comment: 

Would it be "cleaner" to define this as a "size" and that it must be a power-of-two, and apply it to both ipv4 and ipv6 ?  That would reduce those fields to `string` with a common int32 that isn't dependent on v4/v6 ?  It's sort of awkward - what do you think @aojea?

Today:
```
ipv4:
  cidr: 10.0.0.0/16
  perNodeMaskSize: 24
ipv6:
  cidr: dead:beef::/64
  perNodeMaskSize: 120
```

vs could be:
```
perNodeSize: 256
ipv4CIDR: 10.0.0.0/16
ipv6CIDR: dead:beef::/64
```

or even:
```
perNodeBits: 8
ipv4CIDR: 10.0.0.0/16
ipv6CIDR: dead:beef::/64

Reply:
> That would reduce those fields to string with a common int32 that isn't dependent on v4/v6 ?

Agree, I had some doubts, but I think that is better and is aligned with current implementation, the code is v4 and v6 independent, and that makes life to users and developers simpler.

> It's sort of awkward 

yeah, but I think that is more awkward the requirements of having the same size but allowing to set different values, I feel this will save us some time answering issues in the future

> Would it be "cleaner" to define this as a "size" and that it must be a power-of-two, and apply it to both ipv4 and ipv6 ?

I personally think that `perNodeSize`  is easier to understand for non-networking people, or people that don't use to work with network mask

```
perNodeSize: 256
ipv4CIDR: 10.0.0.0/16
ipv6CIDR: dead:beef::/64
```

^^^

Reply:
I agree, having the same size for PerNodeMaskSize is awkward and confusing

I think `perNodeSize` is easier to understand than `perNodeBits`, we just need to be clear that it is documented that `perNodeSize` is not the number of pods that can be scheduled per Node. As per https://github.com/kubernetes/kubernetes/pull/108290#discussion_r837063593 it looks like the ratio is not always fixed

cc: @rahulkjoshi @mskrocki

Reply:
I think perNodeSize is fine -- however, if we collapse the field down to 1 for both ipv4 and ipv6, we make it so we can't give nodes more ipv6 addresses than ipv4. I can see that as a possible request down the line (it would require changes to the scheduler and to the pod spec as well, but it's not infeasible) despite it not being directly possible today.

@mskrocki this is the same issue we're dealing with in multi-network discussions as well -- whether to allow differing numbers of IPs from each network (or in this case IP family)

Reply:
That (different IP sizes) seems like one of those things that will bring endless "fun" implementing and debugging and documenting.  What's the use-case?

Reply:
I think that one of the best decisions we made is to make ipv4 and ipv6 the same, same code and same way of working ... code is sustainable and users don't have to learn nothing new ... I agree with Tim and wait for an use case to justify it

Reply:
SGTM: ping me when ready for re-review

Reply:
2^28 seems like a lot per node? 🙃